cloudsql/sqlserver/database-sql/cloudsql.go

```diff
@@ -0,0 +1,311 @@
+	if _, err = app.db.Exec(`DROP TABLE IF EXISTS votes;`); err != nil {
+		log.Fatalf("DB.Exec: unable to drop votes table: %s", err)
+	}
+	// Create the votes table.
+	_, err = app.db.Exec(`CREATE TABLE votes
+	( vote_id int IDENTITY(1,1) PRIMARY KEY, time_cast DATETIME NOT NULL,
+	candidate CHAR(6) NOT NULL );`)
+	if err != nil {
+		log.Fatalf("DB.Exec: unable to create votes table: %s", err)
+	}
+
+	http.HandleFunc("/", app.indexHandler)
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "8080"
+	}
+
+	log.Printf("Listening on port %s", port)
+	if err := http.ListenAndServe(":"+port, nil); err != nil {
+		log.Fatal(err)
+	}
+
+}
+
+// recentVotes returns a slice of the last 5 votes cast.
+func recentVotes(app *app) ([]vote, error) {
+	var votes []vote
+	rows, err := app.db.Query(`SELECT TOP 5 candidate, time_cast FROM votes ORDER BY time_cast DESC`)
+	if err != nil {
+		return votes, fmt.Errorf("DB.Query: %v", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		nextVote := vote{}
+		err := rows.Scan(&nextVote.Candidate, &nextVote.VoteTime)
+		if err != nil {
+			return votes, fmt.Errorf("Rows.Scan: %v", err)
+		}
+		votes = append(votes, nextVote)
+	}
+	return votes, nil
+}
+
+// currentTotals returns a templateData structure for populating the web page.
+func currentTotals(app *app) (*templateData, error) {
+	// get total votes for each candidate
+	var tabVotes, spaceVotes uint
+	err := app.db.QueryRow(`SELECT count(vote_id) FROM votes WHERE candidate='TABS'`).Scan(&tabVotes)
+	if err != nil {
+		return nil, fmt.Errorf("DB.QueryRow: %v", err)
+	}
+	err = app.db.QueryRow(`SELECT count(vote_id) FROM votes WHERE candidate='SPACES'`).Scan(&spaceVotes)
+	if err != nil {
+		return nil, fmt.Errorf("DB.QueryRow: %v", err)
+	}
+
+	var voteDiffStr string = voteDiff(int(math.Abs(float64(tabVotes) - float64(spaceVotes)))).String()
+
+	latestVotesCast, err := recentVotes(app)
+	if err != nil {
+		return nil, fmt.Errorf("recentVotes: %v", err)
+	}
+
+	var pageData templateData
+	pageData.TabsCount = tabVotes
+	pageData.SpacesCount = spaceVotes
+	pageData.VoteMargin = voteDiffStr
+	pageData.RecentVotes = latestVotesCast
+
+	return &pageData, nil
+}
+
+// showTotals renders an HTML template showing the current vote totals.
+func showTotals(w http.ResponseWriter, r *http.Request, app *app) error {
+	totals, err := currentTotals(app)
+	if err != nil {
+		return fmt.Errorf("currentTotals: %v", err)
+	}
+	err = app.tmpl.Execute(w, totals)
+	if err != nil {
+		return fmt.Errorf("Template.Execute: %v", err)
+	}
+	return nil
+}
+
+// saveVote saves a vote passed as http.Request form data.
+func saveVote(w http.ResponseWriter, r *http.Request, app *app) error {
+	if err := r.ParseForm(); err != nil {
+		return fmt.Errorf("Request.ParseForm: %v", err)
+	}
+
+	var team string
+	if teamprop, ok := r.Form["team"]; ok {
+		team = teamprop[0]
+	} else {
+		return fmt.Errorf("team property missing from form submission")
+	}
```

Comment:
```suggestion
	team := r.FormValue("team")
	if team == "" {
		return fmt.Errorf("team property missing from form submission")
	}
```

Uses the `FormValue` helper and follows https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow.

Reply:
Very cool, I didn't know about the FormValue helper!